src/index/src/inverted_index/create/sort/external_sort.rs

```diff
@@ -0,0 +1,383 @@
+
+            current_memory_usage: 0,
+            memory_usage_threshold,
+        }
+    }
+
+    /// Determines the current data segment based on processed rows
+    fn current_segment_index(&self) -> usize {
+        (self.total_row_count - 1) / self.segment_row_count
+    }
+
+    /// Adds a non-null value to the buffer or updates an existing value's bitmap.
+    /// Returns the memory usage difference of the buffer after the operation.
+    fn push_not_null(&mut self, value: Bytes, offset: usize) -> usize {
+        match self.values_buffer.entry(value) {
+            Entry::Occupied(mut entry) => {
+                let bitmap = entry.get_mut();
+                let old_len = bitmap.as_raw_slice().len();
+                set_bit(bitmap, offset);
+
+                bitmap.as_raw_slice().len() - old_len
+            }
+            Entry::Vacant(entry) => {
+                let key_len = entry.key().len();
+
+                let bitmap = entry.insert(BitVec::default());
+                set_bit(bitmap, offset);
+
+                bitmap.as_raw_slice().len() + key_len
+            }
+        }
+    }
+
+    /// Checks if the in-memory buffer exceeds the threshold and offloads it to external storage if necessary
+    async fn may_dump_buffer(&mut self, memory_diff: usize) -> Result<()> {
+        self.current_memory_usage += memory_diff;
+        if self.current_memory_usage < self.memory_usage_threshold {
+            return Ok(());
+        }
+
+        let values = mem::take(&mut self.values_buffer);
+        let file_id = &format!("{:012}", self.total_row_count);
+
+        let writer = self
+            .temp_file_provider
+            .create(&self.index_name, file_id)
+            .await?;
+        IntermediateWriter::new(writer).write_all(values).await?;
+
+        self.current_memory_usage = 0;
+        Ok(())
+    }
+}
+
+/// Sets or appends a bit at the specified offset within the bitmap
+fn set_bit(bitmap: &mut BitVec, offset: usize) {
+    if offset >= bitmap.len() {
+        bitmap.resize(offset + 1, false);
+    }
+    bitmap.set(offset, true);
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+    use std::iter;
+    use std::sync::Mutex;
+
+    use futures::{AsyncRead, StreamExt};
+    use rand::Rng;
+    use tokio::io::duplex;
+    use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+    use super::*;
+    use crate::inverted_index::create::sort::external_provider::MockExternalTempFileProvider;
+
+    fn generate_random_bytes_option(size: usize) -> Option<Vec<u8>> {
+        let mut rng = rand::thread_rng();
+
+        if rng.gen() {
+            let mut buffer = vec![0u8; size];
+            rng.fill(&mut buffer[..]);
+            Some(buffer)
+        } else {
+            None
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn shuffle_values_and_sorted_result(
+        row_count: usize,
+    ) -> (Vec<Option<Vec<u8>>>, BTreeMap<Option<Vec<u8>>, Vec<usize>>) {
```

Comment:
suggestion(non-blocking): We can add a type alias for `BTreeMap<Option<Vec<u8>>, Vec<usize>>` to help others to understand the contents of the map. e.g. Maybe `ValueRowIds`
```rust
type ValueRowIds = BTreeMap<Option<Vec<u8>>, Vec<usize>>;
```